ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -493,10 +546,20 @@ static int update_object_record_entry(DDTRACE_UNUSED st_data_t *key, st_data_t *
   object_record_update_data *update_data = (object_record_update_data*) data;
   if (existing) {
     object_record *existing_record = (object_record*) (*value);
-    VALUE existing_inspect = object_record_inspect(existing_record);
-    VALUE new_inspect = object_record_inspect(update_data->new_object_record);
-    rb_raise(rb_eRuntimeError, "Object ids are supposed to be unique. We got 2 allocation recordings with "
-        "the same id.\nprevious=%"PRIsVALUE"\nnew=%"PRIsVALUE"\n", existing_inspect, new_inspect);
+    #ifdef HAVE_WORKING_RB_GC_FORCE_RECYCLE
+      // In this case, it's possible for an object id to be re-used. This is not great since we only
+      // look at a small amount of allocation samples which means we may have live object records
+      // being attributed to the wrong stack/labels.
+      // In this particular case, though, we saw the re-used allocation so prefer the new object_record
+      // and cleanup the old one.
+      on_committed_object_record_cleanup(update_data->heap_recorder, existing_record);
```

Comment:
I left a comment about this #3360; I'm somewhat repeating it here in case you missed it -- sorry if you actually did see it but explicitly decided not to do something about it ;)

I somewhat worry that this can hide bugs, e.g. this call to `on_committed_object_record_cleanup` assumes that object id reuse happened because of recycling without actually testing for it.

It's a bit awkward (needs a few changes), but one way of propagating this info would be to introduce a specific struct for `partial_object_record` and then record in `start_heap_allocation_recording` when the bit was set. That way we could get here and do `if (bit was set) { clean record } else { /* fallthrough to raise, it's a bug */ }`.

Reply:
Did it by adding an explicit struct for a `recording`. I think in the end it makes some other intents clearer so I think it's a win overall.